### ISSUE TITLE
NoSuchMethodError when opening Java 21 projects

### ISSUE THREAD

**Issue by BlueGoliath** [apache/netbeans#6373]
### Apache NetBeans version

Apache NetBeans 19 release candidate

### What happened

Netbeans no longer parses Java 21 code and underlines some external dependency class usages as errors. A warning window was created about not being able to parse module-info.java on some project but I'm not able to get it to happen again. All projects build fine via Maven.

nb-javac is uninstalled.

### Did this work correctly in an earlier version?

Apache NetBeans 18

### Operating System

Linux

### JDK

Java 21 x86 built from source

### Apache NetBeans packaging

Apache NetBeans binary zip

### Anything else

```
java.lang.NoSuchMethodError: 'int com.sun.tools.javac.util.Convert.utf2chars(byte[], int, char[], int, int)'
	at org.netbeans.modules.java.source.usages.ClassFileUtil.encodeClassName(ClassFileUtil.java:306)
	at org.netbeans.modules.java.source.usages.ClassFileUtil.encodeClassName(ClassFileUtil.java:218)
	at org.netbeans.modules.java.source.usages.ClassFileUtil.encodeClassNameOrArray(ClassFileUtil.java:211)
	at org.netbeans.modules.java.source.usages.ClassFileUtil.createExecutableDescriptor(ClassFileUtil.java:164)
	at org.netbeans.api.java.source.ElementHandle.createImpl(ElementHandle.java:491)
	at org.netbeans.api.java.source.ElementHandle.create(ElementHandle.java:415)
	at org.netbeans.modules.java.source.parsing.ParameterNameProviderImpl.computeKey(ParameterNameProviderImpl.java:206)
	at org.netbeans.modules.java.source.parsing.ParameterNameProviderImpl.getParameterName(ParameterNameProviderImpl.java:125)
	at org.netbeans.modules.java.source.parsing.ParameterNameProviderImpl$1.invoke(ParameterNameProviderImpl.java:92)
	at jdk.proxy6/jdk.proxy6.$Proxy52.getParameterName(Unknown Source)
	at jdk.compiler/com.sun.tools.javac.code.MissingInfoHandler.getParameterName(MissingInfoHandler.java:64)
	at jdk.compiler/com.sun.tools.javac.code.Symbol$ParamSymbol.getSimpleName(Symbol.java:1882)
	at jdk.compiler/com.sun.tools.javac.code.Symbol$ParamSymbol.getSimpleName(Symbol.java:1867)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.addParameterInlineHint(SemanticHighlighterBase.java:1181)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitIdentifier(SemanticHighlighterBase.java:879)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitIdentifier(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCIdent.accept(JCTree.java:2723)
	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:69)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:1121)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:78)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitMethodInvocation(SemanticHighlighterBase.java:770)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitMethodInvocation(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1832)
	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:69)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:1121)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.source.util.TreeScanner.visitExpressionStatement(TreeScanner.java:503)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitExpressionStatement(SemanticHighlighterBase.java:861)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitExpressionStatement(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1611)
	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:69)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:1121)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:78)
	at jdk.compiler/com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:272)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1104)
	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:69)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:1121)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitMethod(SemanticHighlighterBase.java:934)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitMethod(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:948)
	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:69)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:1121)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:78)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitClass(SemanticHighlighterBase.java:1049)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitClass(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:855)
	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:69)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:1121)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:78)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitCompilationUnit(SemanticHighlighterBase.java:597)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.visitCompilationUnit(SemanticHighlighterBase.java:316)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:623)
	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
	at org.netbeans.api.java.source.support.CancellableTreePathScanner.scan(CancellableTreePathScanner.java:69)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase$DetectorVisitor.scan(SemanticHighlighterBase.java:1121)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase.process(SemanticHighlighterBase.java:200)
	at org.netbeans.modules.java.editor.semantic.SemanticHighlighter.process(SemanticHighlighter.java:56)
	at org.netbeans.modules.java.editor.base.semantic.SemanticHighlighterBase.run(SemanticHighlighterBase.java:154)
	at org.netbeans.modules.parsing.impl.TaskProcessor.callParserResultTask(TaskProcessor.java:561)
	at org.netbeans.modules.parsing.impl.TaskProcessor$RequestPerformer.run(TaskProcessor.java:786)
	at org.openide.util.lookup.Lookups.executeWith(Lookups.java:288)
	at org.netbeans.modules.parsing.impl.TaskProcessor$RequestPerformer.execute(TaskProcessor.java:702)
[catch] at org.netbeans.modules.parsing.impl.TaskProcessor$CompilationJob.run(TaskProcessor.java:663)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at org.openide.util.RequestProcessor$Task.run(RequestProcessor.java:1419)
	at org.netbeans.modules.openide.util.GlobalLookup.execute(GlobalLookup.java:45)
	at org.openide.util.lookup.Lookups.executeWith(Lookups.java:287)
	at org.openide.util.RequestProcessor$Processor.run(RequestProcessor.java:2034)

```

### Are you willing to submit a pull request?

No

**Comment by sdedic** [comment]
I assume NB itself runs on JDK21 - rather that you open a project that targets JDK21.

**Comment by sdedic** [comment]
OK ;) I was mainly curious about the JDK version used to run NB at the error time; it seems  as an incompatibility with JDK21+ (possibly in older NB versions, too) rather than regression between 18 and 19rc.

**Comment by neilcsmith-net** [comment]
This is already fixed in #6206. Unfortunately the javac method signature has changed. The old signature seems to have been in NetBeans code for a while so not sure why you're seeing as a regression (something else causing that code path to be hit perhaps?).

Running without nb-javac is occasionally useful but not supported. It won't be fixed in NB19 and is already fixed in NB20, so closing. Thanks!

**Comment by lahodaj** [comment]
Just for completeness, running with nb-javac uninstalled is meant to be used when the JDK and nb-javac version are "close enough", which typically mean they are from the same JDK release version. I would suggest to try a build from NetBeans master for now if JDK 21 support is desired.
